zio/ndjsonio/typeparser.go

```diff
@@ -0,0 +1,326 @@
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/brimsec/zq/pkg/nano"
+	"github.com/brimsec/zq/zcode"
+	"github.com/brimsec/zq/zio/zeekio"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+	"github.com/buger/jsonparser"
+)
+
+type typeStats struct {
+	BadFormat                int `json:"events_bad_format,omitempty"`
+	BadMetaData              int `json:"events_bad_meta_data,omitempty"`
+	LineTooLong              int `json:"events_line_too_long,omitempty"`
+	FirstBadLine             int `json:"events_first_bad_line,omitempty"`
+	JSONDescriptorNotFound   int `json:"events_json_descriptor_not_found,omitempty"`
+	JSONIncompleteDescriptor int `json:"events_json_incomplete_descriptor,omitempty"`
+	JSONMissingPath          int `json:"events_json_missing_path,omitempty"`
+}
+
+type typeParser struct {
+	zctx        *resolver.Context
+	tr          typeRules
+	defaultPath string
+	stats       *typeStats
+}
+
+var (
+	ErrDescriptorNotFound = errors.New("descriptor not found")
+	ErrMissingPath        = errors.New("missing path field")
+	ErrBadFormat          = errors.New("bad format")
+)
+
+// Information about the correspondence between the flattened structure
+// of a JSON object and its zng representation (which may include
+// nested record fields). The two descriptors here represent the same data
+// in the same order, flatDescriptor describes the data as it appears in
+// JSON, descriptor describes it as it appears in zng values.
+type typeInfo struct {
+	descriptor *zng.TypeRecord
+	flatDesc   *zng.TypeRecord
+	path       []byte
+}
+
+func getUnsafeDefault(data []byte, defaultValue string, key string) (string, error) {
+	val, err := jsonparser.GetUnsafeString(data, key)
+	if err != nil {
+		// This is always a KeyPathNotFoundError, including if the json was invalid.
+		if defaultValue == "" {
+			return "", jsonparser.KeyPathNotFoundError
+		}
+		return defaultValue, nil
+	}
+	return val, nil
+}
+
+func lookupTypeInfo(zctx *resolver.Context, desc *zng.TypeRecord, path string) *typeInfo {
+	flatCols := zeekio.FlattenColumns(desc.Columns)
+	flatDesc := zctx.LookupTypeRecord(flatCols)
+	info := typeInfo{desc, flatDesc, []byte(path)}
+	return &info
+}
+
+// findTypeInfo returns the typeInfo struct matching an input json
+// object.  If no match is found, an error is returned. If defaultPath
+// is non-zero, it is used as the default _path if the object has no
+// such field. (we could at some point make this a bit more generic by
+// passing in a "defaultFieldValues" map... but not needed now).
+func findTypeInfo(zctx *resolver.Context, jobj []byte, tr typeRules, defaultPath string) (*typeInfo, error) {
+	var fieldName, fieldVal, path string
+	for _, r := range tr.rules {
+		if len(r.Match) != 1 {
+			// This will turn into a panic when we do validation upon loading json types
+			return nil, fmt.Errorf("Rule %v with length %d", r, len(r.Match))
+		}
+
+		var name, val string
+		for n, v := range r.Match {
+			name = n
+			val = v
+		}
```

Comment:
Can we make the Rule struct just have Name and Value members instead of the Match map?
The types.json structure could be changed similarly or it could be kept as it currently is and the conversion could be done in an UnmarshalJSON() method on Rule.
(this could happen in a follow-up as it overlaps with validation which you've explicitly mentioned is planned for a follow-up in the comment here)

Reply:
Good idea. I'll just do it on this PR.

Reply:
done